x-pack/plugins/apm/public/components/shared/eql_code_editor/completer.ts

```diff
@@ -0,0 +1,195 @@
+        return [
+          ...withWhitespace(['['], { before: '', after: ' ' }),
+          ...(position.row > 2
+            ? withWhitespace(['until'], { before: '', after: ' [ ' })
+            : []),
+        ];
+
+      case EQLToken.Sequence:
+        return withWhitespace(
+          await this.getExternalSuggestions({
+            type: EQLCodeEditorSuggestionType.Field,
+          }),
+          {
+            after: '\n\t[ ',
+          }
+        );
+
+      case EQLToken.SequenceItemStart:
+        return withWhitespace(
+          [
+            ...(await this.getExternalSuggestions({
+              type: EQLCodeEditorSuggestionType.EventType,
+            })),
+            'any',
+          ],
+          { after: ' where ' }
+        );
+
+      case EQLToken.EventType:
+        return withWhitespace(['where']);
+
+      case EQLToken.Where:
+      case EQLToken.LogicalOperator:
+        return [
+          ...withWhitespace(
+            await this.getExternalSuggestions({
+              type: EQLCodeEditorSuggestionType.Field,
+            })
+          ),
+          ...withWhitespace(['true', 'false'], { after: ' ]\n\t' }),
+        ];
+
+      case EQLToken.BoolCondition:
+        return withWhitespace([']'], { after: '\n\t' });
+
+      case EQLToken.Operator:
+      case EQLToken.InOperator:
+        const field =
+          previousTokens?.find((t) => t.type === EQLToken.Field)?.value ?? '';
+
+        const hasStartedValueLiteral =
+          !!prefix?.trim() || token?.value.trim() === '"';
+
+        return withWhitespace(
+          await this.getExternalSuggestions({
+            type: EQLCodeEditorSuggestionType.Value,
+            field,
+            value: prefix ?? '',
+          }),
+          { before: hasStartedValueLiteral ? '' : ' "', after: '" ' }
+        );
+
+      case EQLToken.Value:
+        return [
+          ...withWhitespace([']'], { after: '\n\t' }),
+          ...withWhitespace(['and', 'or']),
+        ];
+    }
+
+    return [];
+  }
+
+  private async getExternalSuggestions(
+    request: EQLCodeEditorSuggestionRequest
+  ): Promise<EQLCodeEditorSuggestion[]> {
+    if (this.callback) {
+      return this.callback(request);
+    }
+    return [];
+  }
+
+  getCompletions(
+    _: Editor,
+    session: IEditSession,
+    position: { row: number; column: number },
+    prefix: string | undefined,
+    cb: (err: Error | null, suggestions?: EQLCodeEditorSuggestion[]) => void
+  ) {
+    this.getCompletionsAsync(session, position, prefix)
+      .then((suggestions) => {
+        cb(
+          null,
+          suggestions.map((sugg) => {
+            const suggestion =
+              typeof sugg === 'string'
+                ? { value: sugg, score: 1000 }
+                : { score: 1000, ...sugg };
+
+            return suggestion;
+          })
+        );
+      })
+      .catch(cb);
+  }
+
+  setSuggestionCb(cb?: EQLCodeEditorSuggestionCallback) {
+    this.callback = cb;
+  }
+}
```

Comment:
Yeah, so the reason why I haven't spent that much time cleaning this up etc is because I think we're trying to move away from react-ace as a code editor (to Monaco). I couldn't get it to work with Monaco IIRC (this is code from a long time ago). So given that uncertainty, I thought it'd be better to hold off on testing/code sharing etc until we figure out how to make it work with Monaco.